src/context.cc

```diff
@@ -118,29 +122,35 @@ void Context::DelAgent(Agent* m) {
 }
 
 void Context::SchedBuild(Agent* parent, std::string proto_name, int t) {
-  if (t == -1) {
-    t = time() + 1;
+  #pragma omp critical
+  {
+    if (t == -1) {
+      t = time() + 1;
+    }
+    int pid = (parent != NULL) ? parent->id() : -1;
+    ti_->SchedBuild(parent, proto_name, t);
+    NewDatum("BuildSchedule")
+        ->AddVal("ParentId", pid)
+        ->AddVal("Prototype", proto_name)
+        ->AddVal("SchedTime", time())
+        ->AddVal("BuildTime", t)
+        ->Record();
   }
-  int pid = (parent != NULL) ? parent->id() : -1;
-  ti_->SchedBuild(parent, proto_name, t);
-  NewDatum("BuildSchedule")
-      ->AddVal("ParentId", pid)
-      ->AddVal("Prototype", proto_name)
-      ->AddVal("SchedTime", time())
-      ->AddVal("BuildTime", t)
-      ->Record();
 }
 
 void Context::SchedDecom(Agent* m, int t) {
-  if (t == -1) {
-    t = time();
+  #pragma omp critical
```

Comment:
Doesn't this happen once per thread, and therefore generate one database entry per thread?

It may also schedule the decommissioning too many times, and I'm not sure if that matters or not.

Reply:
At each timestep each agent's Tick/Tock is executed by a single thread, so SchedDecom should only be executed once for each agent, right?.  I don't think that this behavior should differ from a serial implementation, apart from perhaps the order of these entries in the DB. There should the the same number of database entries.   Please let me know if I'm missing something about this though, my understanding of an agent's lifecycle might me lacking in some way  

The `#pragma omp critical` block is to ensure that DB writes are not attempted by two threads at the same time.

Reply:
Right... this is called by institutions... makes sense.